### PR TITLE
Add destroy to nmclient in thread

### PR DESCRIPTION
**Pull request by rvykydal**
Follow-up of https://github.com/rhinstaller/anaconda/pull/4322

### CODE REVIEW

File: pyanaconda/modules/network/nm_client.py [hunk 1]
```diff
@@ -72,7 +74,74 @@ def nm_client_in_thread():
...
+    # the shutdown of the NMClient.
+    weak_ref = nm_client.get_context_busy_watcher().weak_ref(_weak_ref_cb)
+    del nm_client
+
+    def _timeout_cb(unused):
+        if not finished:
+            # Somebody else holds a reference to the NMClient and keeps
+            # it alive. We cannot properly clean up.
+            log.debug("[destroy_nm_client]: ERROR: timeout waiting for context "
+                      "busy watcher to be gone")
```
Comment (M4rtinK) [inline pyanaconda/modules/network/nm_client.py:126]: I think this is a bit confusing - maybe "busy watcher to be/needs to be removed" ? Provided I understand this correctly. :)

**Review by M4rtinK — APPROVED** [review]
Looks good to me as far as I can tell. :)

**Comment by rvykydal** [comment]
/kickstart-test network-autoconnections-dhcpall-httpks ifname-httpks network-device-mac-pre

**Comment by rvykydal** [comment]
The kickstart tests are fine, the 3 tests failing in `kickstart-test --testtype network` are flakes.
I tried to re-run them with `kickstart-test network-autoconnections-dhcpall-httpks ifname-httpks network-device-mac-pre` but hit https://github.com/rhinstaller/kickstart-tests/pull/817.
Anyway I am pretty confident the tests are OK, I may re-run the three tests when https://github.com/rhinstaller/kickstart-tests/pull/817 is fixed.

File: pyanaconda/modules/network/nm_client.py (hunk 1 above)
Comment (rvykydal) [reply pyanaconda/modules/network/nm_client.py:126]: Actually the message is:
"timeout waiting for context busy watcher to be gone" meaning it is waiting for "context busy watcher". Unfortunately I split the message based on length, not on the sense, which made it confusing for review I guess.

**Review by VladimirSlavik — APPROVED** [review]
Thank you! Looks good to me.

**Comment by rvykydal** [comment]
/kickstart-test network-autoconnections-dhcpall-httpks ifname-httpks network-device-mac-pre

**Comment by rvykydal** [comment]
@thom311 could you take a look please?
I may update the PR with NM.Client.wait_shutdown_finish when it is released in rawhide.

File: pyanaconda/modules/network/nm_client.py [hunk 2]
```diff
@@ -55,23 +58,91 @@
...
     try:
-        yield get_new_nm_client()
+        nm_client = get_new_nm_client()
     finally:
         mainctx.pop_thread_default()
 
+    # Pass a weak reference to the context so that it does not refer it
+    # when we try to destroy it on context exit.
+    nm_client_ref = weakref.ref(nm_client)
+    yield nm_client_ref
```
Comment (poncovka) [inline pyanaconda/modules/network/nm_client.py:80]: You yield the reference to nm client after `mainctx.pop_thread_default()`, so the nm client will be used outside of the context of the current thread. That doesn't seem right.

File: pyanaconda/modules/network/nm_client.py [hunk 3]
```diff
@@ -55,23 +58,91 @@
...
+
+
+def destroy_nm_client(nm_client_holder):
+    """Destroy NMClient instance.
+
+    NMClient instance will be destroyed if the only refence to it is the one inside
+    the nm_client_holder.
+    Shutting down NMClient properly is tricky.
+    For more information see NetworkManager example examples/python/gi/gmaincontext.py
+    FIXME: use NMClient.wait_shutdown_finish when available
```
Comment (poncovka) [inline pyanaconda/modules/network/nm_client.py:96]: I would prefer to wait for the fix.

**Pull request closed without merging**